app/server/lib/OIDCConfig.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Configuration for OpenID Connect (OIDC), useful for enterprise single-sign-on logins.
+ * A good informative overview of SAML is at https://developer.okta.com/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc
+ * Note:
+ *    SP is "Service Provider", in our case, the Grist application.
+ *    IdP is the "Identity Provider", somewhere users log into, e.g. Okta or Google Apps.
+ *
+ * We also use optional attributes for the user's name, for which we accept any of:
+ *    given_name
+ *    family_name
+ *
+ * Expected environment variables:
+ *    env GRIST_OIDC_SP_HOST=https://<your-domain>
+ *        Host at which our /oidc endpoint will live; identifies our application.
+ *    env GRIST_OIDC_IDP_ISSUER
+ *        The issuer URL for the IdP.
+ *    env GRIST_OIDC_IDP_CLIENT_ID
+ *        The client ID for the application, as registered with the IdP.
+ *    env GRIST_OIDC_IDP_CLIENT_SECRET
+ *        The client secret for the application, as registered with the IdP.
+ *    env GRIST_OIDC_IDP_SCOPES
+ *        The scopes to request from the IdP, as a space-separated list. Defaults to "openid email profile".
+ *    env GRIST_OIDC_IDP_CODE_CHALLENGE_METHOD
+ *        The method to use for code challenge. Defaults to "S256".
+ *
+ * This version of OIDCConfig has been tested with Keycloak OIDC IdP following the instructions
+ * at:
+ *   https://www.keycloak.org/getting-started/getting-started-docker
+ * When running on localhost and http, the settings tested were with:
+ *   - GRIST_OIDC_SP_HOST=http://localhost:8484 (or whatever port you use for Grist)
+ *   - GRIST_OIDC_IDP_ISSUER=http://localhost:8080/realms/myrealm (replace 8080 by the port you use for keycloak)
+ *   - GRIST_OIDC_IDP_CLIENT_ID=myclient
+ *   - GRIST_OIDC_IDP_CLIENT_SECRET=YOUR_SECRET (as set in keycloak)
+ *   - GRIST_OIDC_IDP_SCOPES="openid email profile"
+ */
+
+import * as express from 'express';
+import {GristLoginSystem, GristServer} from './GristServer';
+import { Client, generators, Issuer } from 'openid-client';
+import {Sessions} from './Sessions';
+const CALLBACK_URL = '/oauth2/callback';
+const LOGIN_URL = '/oauth2/login';
+
+export class OIDCConfig {
+  private _client: Client;
+  private _nonce: string;
+  private _codeVerifier: string;
+  private _codeChallenge: string;
+
+  public constructor() {
+    this._nonce = generators.nonce();
```

Comment:
I am pretty sure this is not the right way to initialize a nonce. Instead it should be instantiated each time a user attempts to login.

Probably needs a way to store it. Where?

Reply:
So instead of using `generators.codeVerifier()` included in node-openid-client, which [generates random bytes and stringifies it](https://github.com/panva/node-openid-client/blob/a84d022f195f82ca1c97f8f6b2567ebcef8738c3/lib/helpers/generators.js#L5C32-L5C41), we could take advantage of the permit store, which generates some random string taking advantage of the UUIDv4 algorithm and store it temporarily? I give it a try

PS: I think using the code_verifier alone is enough, neither the `nonce` nor the `state` are useful when dealing with the authorization code flow: https://github.com/panva/node-openid-client/tree/main#authorization-code-flow

Reply:
I tried something, still a bit worried about the implementation regarding the security (the codeVerifier is stored in the permit store the key returned is passed to the identity provider as a GET parameter (as the state), which worries me a bit even if I have no clear idea how an attacker could operate).